Closure Rich Debug Mode

ncteisen opened the pull request:
This change introduces rich debugging of closures.

ncteisen commented:
I am leaving the macro enabled for now so the tests run with it. Once the tests look good, I will disable the rich debug mode and send this out for review

ctiller commented on src/core/lib/iomgr/closure.c:
```diff
@@ -136,29 +168,45 @@ void grpc_closure_run(grpc_exec_ctx *exec_ctx, grpc_closure *c,
+  GPR_TIMER_BEGIN("GRPC_CLOSURE_SCHED", 0);
   if (c != NULL) {
-#ifndef NDEBUG
+#ifdef GRPC_CLOSURE_RICH_DEBUG
```
suggest enabling this with `#ifndef NDEBUG`: it's generally useful enough

ncteisen replied:
Do you think it would be worth it to change all of the custom guards (GRPC_ERROR_REFCOUNT_DEBUG, GRPC_STREAM_REFCOUNT_DEBUG, GRPC_RESOLVER_REFCOUNT_DEBUG, GRPC_LB_POLICY_REFCOUNT_DEBUG) to NDEBUG as well? And maybe additionally turn the into tracers so they can still be turned on and off?

That would solve the problem of the guarded codepaths not being tested on jenkins (to avoid issues like #10691)

ctiller commented:
NDEBUG + tracer variables SGTM

Note that some of our refcount debugging won't pass TSAN (and so will need
some fixups)

On Thu, Apr 20, 2017 at 11:53 AM Noah Eisen <notifications@github.com>
wrote:

> *@ncteisen* commented on this pull request.
> ------------------------------
>
> In src/core/lib/iomgr/closure.c
> <https://github.com/grpc/grpc/pull/10767#discussion_r112533279>:
>
> >    if (c != NULL) {
> -#ifndef NDEBUG
> +#ifdef GRPC_CLOSURE_RICH_DEBUG
>
> Do you think it would be worth it to change all of the custom guards
> (GRPC_ERROR_REFCOUNT_DEBUG, GRPC_STREAM_REFCOUNT_DEBUG,
> GRPC_RESOLVER_REFCOUNT_DEBUG, GRPC_LB_POLICY_REFCOUNT_DEBUG) to NDEBUG as
> well? And maybe additionally turn the into tracers so they can still be
> turned on and off?
>
> That would solve the problem of the guarded codepaths not being tested on
> jenkins (to avoid issues like #10691
> <https://github.com/grpc/grpc/pull/10691>)
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/grpc/grpc/pull/10767#discussion_r112533279>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AJpudSefF6fXS6dc6BgfE8_Zo_Cbmwobks5rx6nEgaJpZM4NDV5l>
> .
>

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
No significant performance differences
```

ncteisen commented:
Last time the tests ran they looked pretty good. I am going to update the environment variables doc with the new tracers tonight.

One question. Is there a reason that any tracers should be enabled for opt builds? I could use this PR to make all tracers dbg only if that makes sense for performance reasons.

ncteisen commented:
I reverted this PR to just deal with closure tracing and moved the bigger tracing overhaul to #10833 to make reviewing it less of a hassle. If you are fine reviewing them as one, feel free to pass over this one

ctiller commented:
This looks good, but I'd like to stall until #10712 and a few friends are merged: they're going to conflict, and some of our timelines are short.

ncteisen commented:
test this please

grpc-testing commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
No significant performance differences
```

ncteisen commented:
I'd love to have this and it's subsequent tracing overhaul PR in before the next grpc_error refcounting crash pops up. I just enabled the macro and ran `make buildtests` so this should be ready to go.

grpc-jenkins commented:
```
No significant performance differences
```

grpc-jenkins commented:
```
Corrupt JSON data (indicates timeout or crash) = {}

Missing files (new benchmark) = {}

No significant performance differences
```

ncteisen commented:
This now built off #11336 for merging conflict convenience

grpc-jenkins commented:
```
Corrupt JSON data (indicates timeout or crash) = {}

Missing files (new benchmark) = {}

Performance differences noted:
Benchmark                                                                                                    atm_add_per_iteration    atm_cas_per_iteration    cpu_time    locks_per_iteration    nows_per_iteration    real_time
-----------------------------------------------------------------------------------------------------------  -----------------------  -----------------------  ----------  ---------------------  --------------------  -----------
BM_PumpStreamClientToServer<InProcessCHTTP2>/0                                                               -7%
BM_PumpStreamClientToServer<InProcessCHTTP2>/1                                                               -7%
BM_PumpStreamClientToServer<InProcessCHTTP2>/4k                                                              -5%
BM_PumpStreamClientToServer<InProcessCHTTP2>/512                                                             -5%
BM_PumpStreamClientToServer<InProcessCHTTP2>/64                                                              -5%
BM_PumpStreamClientToServer<InProcessCHTTP2>/8                                                               -7%
BM_PumpStreamClientToServer<MinInProcessCHTTP2>/0                                                            -7%
BM_PumpStreamClientToServer<MinSockPair>/0                                                                   -7%
BM_PumpStreamClientToServer<MinTCP>/0                                                                        -17%                     -17%                     -10%        -62%                   -99%                  -10%
BM_PumpStreamClientToServer<MinUDS>/0                                                                        -17%                     -17%                                 -49%                   -97%
BM_PumpStreamClientToServer<SockPair>/0                                                                      -7%
BM_PumpStreamClientToServer<SockPair>/1                                                                      -7%
BM_PumpStreamClientToServer<SockPair>/4k                                                                     -5%
BM_PumpStreamClientToServer<SockPair>/512                                                                    -6%
BM_PumpStreamClientToServer<SockPair>/64                                                                     -6%
BM_PumpStreamClientToServer<SockPair>/8                                                                      -7%
BM_PumpStreamClientToServer<TCP>/0                                                                           -18%                     -19%                     -12%        -63%                   -99%                  -12%
BM_PumpStreamClientToServer<TCP>/1                                                                           -9%                      -11%                     -10%        -58%                   -99%                  -11%
BM_PumpStreamClientToServer<TCP>/256k                                                                                                 -4%                                  -29%                   -6%
BM_PumpStreamClientToServer<TCP>/2M                                                                                                                                        -14%
BM_PumpStreamClientToServer<TCP>/32k                                                                         -15%                     -18%                                 -51%                   -50%
BM_PumpStreamClientToServer<TCP>/4k                                                                          -20%                     -24%                                 -64%                   -90%
BM_PumpStreamClientToServer<TCP>/512                                                                         -17%                     -22%                                 -64%                   -98%
BM_PumpStreamClientToServer<TCP>/64                                                                          -12%                     -17%                                 -62%                   -99%
BM_PumpStreamClientToServer<TCP>/8                                                                           -17%                     -18%                     -6%         -62%                   -99%                  -5%
BM_PumpStreamClientToServer<UDS>/0                                                                           -17%                     -17%                                 -49%                   -97%
BM_PumpStreamClientToServer<UDS>/1                                                                           -17%                     -17%                                 -49%                   -97%
BM_PumpStreamClientToServer<UDS>/256k                                                                                                                                      -15%
BM_PumpStreamClientToServer<UDS>/32k                                                                         -8%                      -7%                                  -28%                   -17%
BM_PumpStreamClientToServer<UDS>/4k                                                                          -14%                     -14%                                 -46%                   -77%
BM_PumpStreamClientToServer<UDS>/512                                                                         -14%                     -16%                                 -49%                   -94%
BM_PumpStreamClientToServer<UDS>/64                                                                          -14%                     -17%                                 -49%                   -97%
BM_PumpStreamClientToServer<UDS>/8                                                                           -17%                     -17%                                 -49%                   -97%
BM_PumpStreamServerToClient<InProcessCHTTP2>/0                                                               -7%
BM_PumpStreamServerToClient<InProcessCHTTP2>/1                                                               -7%
BM_PumpStreamServerToClient<InProcessCHTTP2>/4k                                                              -5%
BM_PumpStreamServerToClient<InProcessCHTTP2>/512                                                             -5%
BM_PumpStreamServerToClient<InProcessCHTTP2>/64                                                              -5%
BM_PumpStreamServerToClient<InProcessCHTTP2>/8                                                               -7%
BM_PumpStreamServerToClient<MinInProcessCHTTP2>/0                                                            -7%
BM_PumpStreamServerToClient<MinSockPair>/0                                                                   -7%
BM_PumpStreamServerToClient<MinTCP>/0                                                                        -14%                     -15%                     -14%        -61%                   -99%                  -14%
BM_PumpStreamServerToClient<MinUDS>/0                                                                        -17%                     -17%                     -5%         -49%                   -97%                  -5%
BM_PumpStreamServerToClient<SockPair>/0                                                                      -7%
BM_PumpStreamServerToClient<SockPair>/1                                                                      -7%
BM_PumpStreamServerToClient<SockPair>/4k                                                                     -6%
BM_PumpStreamServerToClient<SockPair>/512                                                                    -6%
BM_PumpStreamServerToClient<SockPair>/64                                                                     -6%
BM_PumpStreamServerToClient<SockPair>/8                                                                      -7%
BM_PumpStreamServerToClient<TCP>/0                                                                           -16%                     -17%                     -16%        -62%                   -99%                  -15%
BM_PumpStreamServerToClient<TCP>/1                                                                           -10%                     -12%                     -12%        -59%                   -99%                  -12%
BM_PumpStreamServerToClient<TCP>/256k                                                                                                                                      -30%
BM_PumpStreamServerToClient<TCP>/32k                                                                         -13%                     -16%                                 -50%                   -44%
BM_PumpStreamServerToClient<TCP>/4k                                                                          -20%                     -24%                                 -64%                   -90%
BM_PumpStreamServerToClient<TCP>/512                                                                         -17%                     -22%                                 -64%                   -98%
BM_PumpStreamServerToClient<TCP>/64                                                                          -11%                     -17%                                 -62%                   -99%
BM_PumpStreamServerToClient<TCP>/8                                                                           -14%                     -14%                                 -61%                   -99%
BM_PumpStreamServerToClient<UDS>/0                                                                           -17%                     -17%                     -4%         -49%                   -97%                  -4%
BM_PumpStreamServerToClient<UDS>/1                                                                           -17%                     -17%                     -7%         -49%                   -97%                  -7%
BM_PumpStreamServerToClient<UDS>/256k                                                                                                                                      -18%                   -4%
BM_PumpStreamServerToClient<UDS>/32k                                                                         -9%                      -10%                                 -32%                   -23%
BM_PumpStreamServerToClient<UDS>/4k                                                                          -14%                     -16%                                 -46%                   -78%
BM_PumpStreamServerToClient<UDS>/512                                                                         -14%                     -17%                                 -49%                   -94%
BM_PumpStreamServerToClient<UDS>/64                                                                          -14%                     -17%                                 -49%                   -97%
BM_PumpStreamServerToClient<UDS>/8                                                                           -17%                     -17%                                 -49%                   -97%
BM_StreamingPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/0/0                                          -4%
BM_StreamingPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/0/1                                          -5%
BM_StreamingPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/0/2                                          -6%
BM_StreamingPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/1/1                                          -5%
BM_StreamingPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/1/2                                          -6%
BM_StreamingPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/1                                         -4%
BM_StreamingPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/2                                         -5%
BM_StreamingPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/512/1                                        -5%
BM_StreamingPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/512/2                                        -5%
BM_StreamingPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/64/1                                         -5%
BM_StreamingPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/64/2                                         -5%
BM_StreamingPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/8/1                                          -5%
BM_StreamingPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/8/2                                          -5%
BM_StreamingPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/0                                       -4%
BM_StreamingPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/1                                       -5%
BM_StreamingPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/2                                       -6%
BM_StreamingPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/1/1                                       -5%
BM_StreamingPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/1/2                                       -6%
BM_StreamingPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/1                                      -4%
BM_StreamingPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/2                                      -4%
BM_StreamingPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/512/1                                     -5%
BM_StreamingPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/512/2                                     -5%
BM_StreamingPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/64/1                                      -5%
BM_StreamingPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/64/2                                      -5%
BM_StreamingPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/8/1                                       -5%
BM_StreamingPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/8/2                                       -6%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/0/0                                                   -10%                     -7%                                  -17%                   -8%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/0/1                                                   -13%                     -8%                                  -21%                   -22%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/0/2                                                   -14%                     -9%                                  -23%                   -20%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/1/1                                                   -12%                     -8%                                  -21%                   -22%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/1/2                                                   -13%                     -9%                                  -23%                   -20%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/256k/1                                                -7%                      -4%                                  -17%                   -12%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/256k/2                                                -7%                      -4%                                  -18%                   -9%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/2M/1                                                                                                                -12%                   -9%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/2M/2                                                                                                                -11%                   -4%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/32k/1                                                 -10%                     -7%                                  -19%                   -15%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/32k/2                                                 -9%                      -6%                                  -20%                   -8%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/4k/1                                                  -11%                     -8%                                  -20%                   -19%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/4k/2                                                  -11%                     -8%                                  -21%                   -16%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/512/1                                                 -11%                     -8%                                  -20%                   -21%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/512/2                                                 -12%                     -9%                                  -22%                   -18%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/64/1                                                  -11%                     -8%                                  -21%                   -21%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/64/2                                                  -12%                     -9%                                  -22%                   -19%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/8/1                                                   -12%                     -8%                                  -21%                   -22%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/8/2                                                   -13%                     -9%                                  -23%                   -20%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/0/0                                                      -10%                     -6%                                  -17%                   -8%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/0/1                                                      -13%                     -8%                                  -21%                   -22%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/0/2                                                      -14%                     -9%                                  -23%                   -20%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/1/1                                                      -12%                     -8%                                  -21%                   -22%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/1/2                                                      -13%                     -9%                                  -23%                   -20%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/256k/1                                                   -7%                      -4%                                  -17%                   -10%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/256k/2                                                   -5%                                                           -17%                   -6%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/2M/1                                                                                                                   -13%                   -9%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/2M/2                                                                                                                   -13%                   -6%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/32k/1                                                    -10%                     -6%                                  -19%                   -15%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/32k/2                                                    -9%                      -5%                                  -19%                   -7%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/4k/1                                                     -11%                     -7%                                  -20%                   -19%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/4k/2                                                     -11%                     -8%                                  -21%                   -16%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/512/1                                                    -11%                     -8%                                  -20%                   -21%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/512/2                                                    -12%                     -8%                                  -22%                   -18%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/64/1                                                     -11%                     -8%                                  -21%                   -21%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/64/2                                                     -12%                     -8%                                  -22%                   -19%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/8/1                                                      -12%                     -8%                                  -21%                   -22%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/8/2                                                      -13%                     -9%                                  -23%                   -20%
BM_StreamingPingPongMsgs<InProcessCHTTP2, NoOpMutator, NoOpMutator>/0                                        -7%
BM_StreamingPingPongMsgs<InProcessCHTTP2, NoOpMutator, NoOpMutator>/1                                        -7%
BM_StreamingPingPongMsgs<InProcessCHTTP2, NoOpMutator, NoOpMutator>/4k                                       -4%
BM_StreamingPingPongMsgs<InProcessCHTTP2, NoOpMutator, NoOpMutator>/512                                      -5%
BM_StreamingPingPongMsgs<InProcessCHTTP2, NoOpMutator, NoOpMutator>/64                                       -5%
BM_StreamingPingPongMsgs<InProcessCHTTP2, NoOpMutator, NoOpMutator>/8                                        -7%
BM_StreamingPingPongMsgs<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0                                     -7%
BM_StreamingPingPongMsgs<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/1                                     -7%
BM_StreamingPingPongMsgs<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/4k                                    -4%
BM_StreamingPingPongMsgs<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/512                                   -5%
BM_StreamingPingPongMsgs<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/64                                    -5%
BM_StreamingPingPongMsgs<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/8                                     -7%
BM_StreamingPingPongMsgs<MinTCP, NoOpMutator, NoOpMutator>/0                                                 -17%                     -11%                                 -27%                   -13%
BM_StreamingPingPongMsgs<MinTCP, NoOpMutator, NoOpMutator>/1                                                 -17%                     -10%                                 -27%                   -13%
BM_StreamingPingPongMsgs<MinTCP, NoOpMutator, NoOpMutator>/256k                                              -7%                      -5%                                  -20%                   -5%
BM_StreamingPingPongMsgs<MinTCP, NoOpMutator, NoOpMutator>/2M                                                                                                              -6%
BM_StreamingPingPongMsgs<MinTCP, NoOpMutator, NoOpMutator>/32k                                               -10%                     -7%                                  -20%
BM_StreamingPingPongMsgs<MinTCP, NoOpMutator, NoOpMutator>/4k                                                -13%                     -9%                                  -23%                   -9%
BM_StreamingPingPongMsgs<MinTCP, NoOpMutator, NoOpMutator>/512                                               -13%                     -10%                                 -24%                   -11%
BM_StreamingPingPongMsgs<MinTCP, NoOpMutator, NoOpMutator>/64                                                -15%                     -10%                                 -25%                   -12%
BM_StreamingPingPongMsgs<MinTCP, NoOpMutator, NoOpMutator>/8                                                 -17%                     -10%                                 -27%                   -13%
BM_StreamingPingPongMsgs<TCP, NoOpMutator, NoOpMutator>/0                                                    -17%                     -11%                                 -27%                   -13%
BM_StreamingPingPongMsgs<TCP, NoOpMutator, NoOpMutator>/1                                                    -17%                     -10%                                 -27%                   -13%
BM_StreamingPingPongMsgs<TCP, NoOpMutator, NoOpMutator>/256k                                                 -5%                                                           -19%
BM_StreamingPingPongMsgs<TCP, NoOpMutator, NoOpMutator>/2M                                                                                                                 -6%
BM_StreamingPingPongMsgs<TCP, NoOpMutator, NoOpMutator>/32k                                                  -10%                     -7%                                  -20%
BM_StreamingPingPongMsgs<TCP, NoOpMutator, NoOpMutator>/4k                                                   -12%                     -9%                                  -23%                   -9%
BM_StreamingPingPongMsgs<TCP, NoOpMutator, NoOpMutator>/512                                                  -13%                     -10%                                 -24%                   -11%
BM_StreamingPingPongMsgs<TCP, NoOpMutator, NoOpMutator>/64                                                   -15%                     -10%                                 -25%                   -13%
BM_StreamingPingPongMsgs<TCP, NoOpMutator, NoOpMutator>/8                                                    -17%                     -10%                                 -27%                   -13%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/0/0/0                       -4%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/0/0/1                       -4%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/0/1/0                       -5%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/0/1/1                       -5%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/0/2/0                       -6%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/0/2/1                       -5%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/1/1/0                       -5%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/1/1/1                       -5%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/1/2/0                       -6%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/1/2/1                       -5%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/1/0                      -4%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/1/1                      -4%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/2/0                      -4%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/2/1                      -4%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/512/1/0                     -5%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/512/1/1                     -4%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/512/2/0                     -5%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/512/2/1                     -5%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/64/1/0                      -5%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/64/1/1                      -4%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/64/2/0                      -5%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/64/2/1                      -5%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/8/1/0                       -5%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/8/1/1                       -5%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/8/2/0                       -6%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/8/2/1                       -5%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/0/0                    -4%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/0/1                    -4%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/1/0                    -5%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/1/1                    -5%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/2/0                    -6%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/2/1                    -5%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/1/1/0                    -5%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/1/1/1                    -5%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/1/2/0                    -6%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/1/2/1                    -5%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/1/0                   -4%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/1/1                   -4%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/2/0                   -4%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/2/1                   -4%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/512/1/0                  -5%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/512/1/1                  -4%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/512/2/0                  -5%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/512/2/1                  -5%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/64/1/0                   -5%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/64/1/1                   -4%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/64/2/0                   -5%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/64/2/1                   -5%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/8/1/0                    -5%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/8/1/1                    -5%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/8/2/0                    -6%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/8/2/1                    -5%
BM_UnaryPingPong<InProcessCHTTP2, Client_AddMetadata<RandomAsciiMetadata<100>, 1>, NoOpMutator>/0/0          -4%
BM_UnaryPingPong<InProcessCHTTP2, Client_AddMetadata<RandomAsciiMetadata<10>, 1>, NoOpMutator>/0/0           -4%
BM_UnaryPingPong<InProcessCHTTP2, Client_AddMetadata<RandomAsciiMetadata<31>, 1>, NoOpMutator>/0/0           -4%
BM_UnaryPingPong<InProcessCHTTP2, Client_AddMetadata<RandomBinaryMetadata<100>, 1>, NoOpMutator>/0/0         -4%
BM_UnaryPingPong<InProcessCHTTP2, Client_AddMetadata<RandomBinaryMetadata<100>, 2>, NoOpMutator>/0/0         -4%
BM_UnaryPingPong<InProcessCHTTP2, Client_AddMetadata<RandomBinaryMetadata<10>, 1>, NoOpMutator>/0/0          -4%
BM_UnaryPingPong<InProcessCHTTP2, Client_AddMetadata<RandomBinaryMetadata<10>, 2>, NoOpMutator>/0/0          -4%
BM_UnaryPingPong<InProcessCHTTP2, Client_AddMetadata<RandomBinaryMetadata<31>, 1>, NoOpMutator>/0/0          -4%
BM_UnaryPingPong<InProcessCHTTP2, Client_AddMetadata<RandomBinaryMetadata<31>, 2>, NoOpMutator>/0/0          -4%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/0/0                                              -5%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/0/1                                              -5%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/0/4k                                             -4%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/0/512                                            -4%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/0/64                                             -4%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/0/8                                              -5%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/1/0                                              -5%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/1/1                                              -5%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/0                                             -4%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/4k                                            -4%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/512/0                                            -4%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/512/512                                          -4%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/64/0                                             -4%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/64/64                                            -4%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/8/0                                              -5%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/8/8                                              -5%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, Server_AddInitialMetadata<RandomAsciiMetadata<100>, 1>>/0/0   -4%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, Server_AddInitialMetadata<RandomAsciiMetadata<10>, 1>>/0/0    -4%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, Server_AddInitialMetadata<RandomAsciiMetadata<31>, 1>>/0/0    -4%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, Server_AddInitialMetadata<RandomBinaryMetadata<100>, 1>>/0/0  -4%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, Server_AddInitialMetadata<RandomBinaryMetadata<10>, 1>>/0/0   -4%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, Server_AddInitialMetadata<RandomBinaryMetadata<31>, 1>>/0/0   -4%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/0                                           -5%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/1                                           -5%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/4k                                          -4%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/512                                         -4%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/64                                          -4%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/8                                           -5%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/1/0                                           -5%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/1/1                                           -5%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/0                                          -4%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/4k                                         -4%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/512/0                                         -4%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/512/512                                       -4%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/64/0                                          -4%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/64/64                                         -4%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/8/0                                           -5%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/8/8                                           -5%
BM_UnaryPingPong<MinSockPair, NoOpMutator, NoOpMutator>/0/0                                                  -4%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/0/0                                                       -10%                     -8%                                  -19%                   -8%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/0/1                                                       -10%                     -7%                                  -19%                   -8%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/0/256k                                                    -6%                                                           -16%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/0/2M                                                                                                                    -7%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/0/32k                                                     -9%                      -7%                                  -18%                   -4%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/0/4k                                                      -9%                      -7%                                  -18%                   -7%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/0/512                                                     -10%                     -7%                                  -18%                   -8%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/0/64                                                      -10%                     -7%                                  -18%                   -8%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/0/8                                                       -10%                     -8%                                  -19%                   -8%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/1/0                                                       -10%                     -8%                                  -19%                   -8%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/1/1                                                       -10%                     -8%                                  -19%                   -8%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/256k/0                                                    -7%                      -4%                                  -15%                   -5%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/256k/256k                                                 -5%                                                           -14%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/32k/0                                                     -8%                      -6%                                  -15%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/32k/32k                                                   -7%                      -5%                                  -15%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/4k/0                                                      -9%                      -7%                                  -18%                   -7%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/4k/4k                                                     -9%                      -7%                                  -17%                   -6%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/512/0                                                     -10%                     -7%                                  -18%                   -8%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/512/512                                                   -9%                      -7%                                  -17%                   -8%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/64/0                                                      -10%                     -7%                                  -18%                   -8%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/64/64                                                     -9%                      -7%                                  -18%                   -8%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/8/0                                                       -10%                     -7%                                  -19%                   -8%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/8/8                                                       -10%                     -7%                                  -19%                   -8%
BM_UnaryPingPong<MinUDS, NoOpMutator, NoOpMutator>/0/0                                                       -6%                                                           -5%                    -9%
BM_UnaryPingPong<SockPair, NoOpMutator, NoOpMutator>/0/0                                                     -4%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/0                                                          -10%                     -7%                                  -19%                   -8%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/1                                                          -10%                     -7%                                  -19%                   -8%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/256k                                                       -7%                      -4%                                  -17%                   -5%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/2M                                                                                                                       -7%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/32k                                                        -8%                      -5%                                  -18%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/4k                                                         -9%                      -6%                                  -18%                   -7%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/512                                                        -10%                     -7%                                  -18%                   -8%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/64                                                         -10%                     -7%                                  -18%                   -8%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/0/8                                                          -10%                     -7%                                  -19%                   -8%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/1/0                                                          -10%                     -7%                                  -19%                   -8%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/1/1                                                          -10%                     -7%                                  -19%                   -8%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/256k/0                                                       -6%                                                           -14%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/256k/256k                                                    -4%                                                           -13%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/32k/0                                                        -8%                      -5%                                  -15%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/32k/32k                                                      -7%                      -4%                                  -15%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/4k/0                                                         -9%                      -7%                                  -18%                   -7%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/4k/4k                                                        -9%                      -6%                                  -17%                   -6%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/512/0                                                        -10%                     -7%                                  -18%                   -8%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/512/512                                                      -9%                      -7%                                  -18%                   -8%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/64/0                                                         -10%                     -7%                                  -18%                   -8%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/64/64                                                        -9%                      -7%                                  -18%                   -8%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/8/0                                                          -10%                     -7%                                  -19%                   -8%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/8/8                                                          -10%                     -7%                                  -19%                   -8%
BM_UnaryPingPong<UDS, NoOpMutator, NoOpMutator>/0/0                                                          -6%                                                           -5%                    -9%
```

grpc-jenkins commented:
```
Corrupt JSON data (indicates timeout or crash) = {}

Missing files (new benchmark) = {}

No significant performance differences
```

ncteisen commented:
Mac failures explained by: #11223 
Linux sanity fixed in: #11443 

Gonna run the whole test suite one more time to see if tsan and asan are flakes or a real issue. They shouldn't be affected as this is all macros

ncteisen commented:
jenkins: test this please

grpc-jenkins commented:
```
Corrupt JSON data (indicates timeout or crash) = {}

Missing files (new benchmark) = {}

No significant performance differences
```

ncteisen commented:
Mac and Linux failures previously stated.
The asan failure is a flake (9/100) runs failed locally
